### PR TITLE
[bitnami/zookeeper] add persistence.existingClaim

### LINKED ISSUE #2185: [bitnami/zookeeper] support persistence.existingClaim
**Which chart**:
zookeeper 5.7.2

**Is your feature request related to a problem? Please describe.**
The kafka chart supports persistence.existingClaim as an option, but without zookeeper supporting the same option, it can't be deployed the same way. 

**Describe the solution you'd like**
Zookeeper could add the same mechanism (use an existing PersistentStorageClaim) with 'persistence.existingClaim') as the kafka chart does.

**Describe alternatives you've considered**
On a system that can't provide a StorageClass, zookeeper can only run with emptyDir storage (by setting persistence.enabled=false).

### PR DESCRIPTION
**Pull request by horihel**
**Description of the change**

This change will add the value `persistence.useExisting` to support using an existing pvc 

The code for this was shamelessly copied over from bitnami/kafka.

**Benefits**

zookeeper will now support the same features wrt persistence as kafka.

**Possible drawbacks**

none AFAIK

**Applicable issues**

  - fixes #2185

**Additional information**

I'm not sure how the chart version should be bumped for this. I only bumped bugfix, please advise.

**Checklist** 
- [X] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/).
- [X] Variables are documented in the README.md
- [X] Title of the PR starts with chart name (e.g. `[bitnami/chart]`)
- [X] If the chart contains a `values-production.yaml` apart from `values.yaml`, ensure that you implement the changes in both files

:warning: Keep in mind that if you want to make changes to the kubeapps chart, please implement them in the [kubeapps repository](https://github.com/kubeapps/kubeapps/tree/master/chart/kubeapps). This is only a synchronized mirror.

### CODE REVIEW

File: bitnami/zookeeper/Chart.yaml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 apiVersion: v1
 name: zookeeper
-version: 5.7.2
+version: 5.7.3
```
Comment (miguelaeh) [inline bitnami/zookeeper/Chart.yaml:3]: As you are adding a new feature this should be a minor change

File: bitnami/zookeeper/README.md [hunk 2]
```diff
@@ -103,6 +103,7 @@ The following tables lists the configurable parameters of the ZooKeeper chart an
 | `securityContext.fsGroup`              | Group ID for the container (ZooKeeper master pod)                                                                                                         | `1001`                                                       |  |
 | `securityContext.runAsUser`            | User ID for the container (ZooKeeper master pod)                                                                                                          | `1001`                                                       |  |
 | `persistence.enabled`                  | Enable persistence using PVC                                                                                                                              | `true`                                                       |  |
+| `persistence.existingClaim` | Provide an existing `PersistentVolumeClaim`, the value is evaluated as a template      | `nil`                         |
```
Comment (miguelaeh) [inline bitnami/zookeeper/README.md:106]: Please preserve the indentation as the rest of the table

File: bitnami/zookeeper/templates/statefulset.yaml [hunk 3]
```diff
@@ -184,11 +184,14 @@ spec:
           configMap:
             name: {{ template "zookeeper.fullname" . }}
         {{- end }}
-        {{- if not .Values.persistence.enabled }}
+{{- if not .Values.persistence.enabled }}
```
Comment (miguelaeh) [inline bitnami/zookeeper/templates/statefulset.yaml:187]: Please indent the templates at the same level as the code. This applies to the rest of the file

File: bitnami/zookeeper/templates/statefulset.yaml [hunk 4]
```diff
@@ -184,11 +184,14 @@ spec:
...
         - name: data
           emptyDir: {}
-        {{- end }}
-  {{- if .Values.persistence.enabled }}
+{{- else if .Values.persistence.existingClaim }}
```
Comment (miguelaeh) [inline bitnami/zookeeper/templates/statefulset.yaml:190]: I would revert the logic to be more understandable. For example:
```

       {{- if and .Values.persistence.enabled .Values.persistence.existingClaim }}
        - name: data
          persistentVolumeClaim:
            claimName: {{ .Values.persistence.existingClaim }}
       {{- else }}
       - name: data
          emptyDir: {}
       {{- end }}

File: bitnami/zookeeper/values-production.yaml [hunk 5]
```diff
@@ -185,6 +185,12 @@ securityContext:
 ##
 persistence:
   enabled: true
+  ## A manually managed Persistent Volume and Claim
+  ## If defined, PVC must be created manually before volume will be bound
+  ## The value is evaluated as a template
+  ##
+  
+  # existingClaim:
```
Comment (miguelaeh) [inline bitnami/zookeeper/values-production.yaml:193]: Add an empty line after this

File: bitnami/zookeeper/values-production.yaml [hunk 6]
```diff
@@ -185,6 +185,12 @@ securityContext:
 ##
 persistence:
   enabled: true
+  ## A manually managed Persistent Volume and Claim
+  ## If defined, PVC must be created manually before volume will be bound
+  ## The value is evaluated as a template
+  ##
+
```
Comment (miguelaeh) [inline bitnami/zookeeper/values-production.yaml:192]: Remove this empty line

File: bitnami/zookeeper/Chart.yaml (hunk 1 above)
Comment (horihel) [reply bitnami/zookeeper/Chart.yaml]: fixed

File: bitnami/zookeeper/templates/statefulset.yaml (hunk 4 above)
Comment (horihel) [reply bitnami/zookeeper/templates/statefulset.yaml:190]: logic is reverted now - please mind that the original logic was copied from bitnami/kafka - so they differ now

**Comment by horihel** [comment]
Hi @miguelaeh 
Thank you very much for reviewing this.
I think I fixed the change requests properly, including the logic change in statefulset.yaml

**Comment by horihel** [comment]
oh wait- we might have a problem now. If persistence.enabled=true and no existingClaim is specified, it will assign an emptyDir.

This seems to get overridden by the VolumeClaimTemplate, but it's still not correct. I'll revise.

**Comment by horihel** [comment]
now it should be correct.

**Comment by horihel** [comment]
sorry, by trying to resolve a merge conflict i totally messed up that branch. I'll close this and do a new PR.
sorry

**Pull request closed without merging**